File: Algo/leetcode/algorithms/HashMap/383.py
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        magaDict = {}
        for m in magazine:
            if m in magaDict:
                magaDict[m] = magaDict[m] + 1
            else:
                magaDict[m] = 1
        
        for r in ransomNote:
            if r not in magaDict:
                return False
            else:
                t = magaDict[r]
                if t - 1 < 0:
                    return False
                magaDict[r] = t -1
        return True

File: Algo/leetcode/algorithms/HashMap/383.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> magaDict = new HashMap<Character,Integer>();
        for (Character m : magazine.toCharArray()) {
            Integer cnt = magaDict.get(m);
            if (cnt == null) {
                cnt = 0;
            }
            magaDict.put(m, cnt + 1);
        }
        for (Character r : ransomNote.toCharArray()){
            Integer cnt = magaDict.get(r);
            if (cnt == null || cnt <= 0) {
                return false;
            } else {
                magaDict.put(r, cnt - 1);
            }
        }
        return true;
    }
}